File: backend-dotnet/src/Infrastructure/App.Template.Infrastructure/Services/FileStorageService.cs
using AppTemplate.Application.Interfaces;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;

namespace AppTemplate.Infrastructure.Services;

public class FileStorageService : IFileStorageService
{
    private readonly string _storagePath;
    private readonly ILogger<FileStorageService> _logger;

    public FileStorageService(IConfiguration configuration, ILogger<FileStorageService> logger)
    {
        _storagePath = configuration["FileStorage:Path"] ?? Path.Combine(Directory.GetCurrentDirectory(), "uploads");
        _logger = logger;

        // Ensure the upload directory exists
        if (!Directory.Exists(_storagePath))
        {
            Directory.CreateDirectory(_storagePath);
            _logger.LogInformation("Created upload directory: {Path}", _storagePath);
        }
    }

    public async Task<(string FileName, string StoragePath)> SaveFileAsync(
        Stream fileStream,
        string originalFileName,
        string contentType,
        CancellationToken cancellationToken = default)
    {
        // Generate unique filename
        var fileExtension = Path.GetExtension(originalFileName);
        var uniqueFileName = $"{Guid.NewGuid()}{fileExtension}";

        // Create year/month subfolder structure
        var subFolder = DateTime.UtcNow.ToString("yyyy/MM");
        var folderPath = Path.Combine(_storagePath, subFolder);

        if (!Directory.Exists(folderPath))
        {
            Directory.CreateDirectory(folderPath);
        }

        var fullPath = Path.Combine(folderPath, uniqueFileName);
        var storagePath = Path.Combine(subFolder, uniqueFileName).Replace("\\", "/");

        await using var fileStreamOut = new FileStream(fullPath, FileMode.Create);
        await fileStream.CopyToAsync(fileStreamOut, cancellationToken);

        _logger.LogInformation("File saved: {FileName} -> {StoragePath}", originalFileName, storagePath);

        return (uniqueFileName, storagePath);
    }

    public async Task<Stream?> GetFileAsync(string storagePath, CancellationToken cancellationToken = default)
    {
        var fullPath = Path.Combine(_storagePath, storagePath.Replace("/", Path.DirectorySeparatorChar.ToString()));

        if (!File.Exists(fullPath))
        {
            _logger.LogWarning("File not found: {StoragePath}", storagePath);
            return null;
        }

        var memoryStream = new MemoryStream();
        await using var fileStream = new FileStream(fullPath, FileMode.Open, FileAccess.Read);
        await fileStream.CopyToAsync(memoryStream, cancellationToken);
        memoryStream.Position = 0;

        return memoryStream;
    }

    public Task DeleteFileAsync(string storagePath, CancellationToken cancellationToken = default)
    {
        var fullPath = Path.Combine(_storagePath, storagePath.Replace("/", Path.DirectorySeparatorChar.ToString()));

        if (File.Exists(fullPath))
        {
            File.Delete(fullPath);
            _logger.LogInformation("File deleted: {StoragePath}", storagePath);
        }
        else
        {
            _logger.LogWarning("File not found for deletion: {StoragePath}", storagePath);
        }

        return Task.CompletedTask;
    }

    public bool FileExists(string storagePath)
    {
        var fullPath = Path.Combine(_storagePath, storagePath.Replace("/", Path.DirectorySeparatorChar.ToString()));
        return File.Exists(fullPath);
    }
}

File: backend/nestjs/clean-architecture/full/src/modules/file-management/infrastructure/services/file-storage.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Readable } from 'stream';
import * as fs from 'fs';
import * as path from 'path';
import { v4 as uuidv4 } from 'uuid';
import {
  IFileStorageService,
  FileStorageResult,
} from '../../domain/interfaces/file-storage.service.interface';

@Injectable()
export class FileStorageService implements IFileStorageService {
  private readonly logger = new Logger(FileStorageService.name);
  private readonly storagePath: string;

  constructor(private readonly configService: ConfigService) {
    this.storagePath = this.configService.get<string>('FILE_STORAGE_PATH') || './uploads';

    // Ensure the upload directory exists
    if (!fs.existsSync(this.storagePath)) {
      fs.mkdirSync(this.storagePath, { recursive: true });
      this.logger.log(`Created upload directory: ${this.storagePath}`);
    }
  }

  async saveFile(
    fileBuffer: Buffer,
    originalFileName: string,
    contentType: string,
  ): Promise<FileStorageResult> {
    // Generate unique filename
    const ext = path.extname(originalFileName);
    const uniqueFileName = `${uuidv4()}${ext}`;

    // Create year/month subfolder structure
    const now = new Date();
    const subFolder = `${now.getFullYear()}/${String(now.getMonth() + 1).padStart(2, '0')}`;
    const folderPath = path.join(this.storagePath, subFolder);

    if (!fs.existsSync(folderPath)) {
      fs.mkdirSync(folderPath, { recursive: true });
    }

    const filePath = path.join(folderPath, uniqueFileName);
    const storagePath = `${subFolder}/${uniqueFileName}`;

    await fs.promises.writeFile(filePath, fileBuffer);

    this.logger.log(`File saved: ${originalFileName} -> ${storagePath}`);

    return { fileName: uniqueFileName, storagePath };
  }

  async getFile(storagePath: string): Promise<Readable | null> {
    const fullPath = path.join(this.storagePath, storagePath);

    if (!fs.existsSync(fullPath)) {
      this.logger.warn(`File not found: ${storagePath}`);
      return null;
    }

    return fs.createReadStream(fullPath);
  }

  async deleteFile(storagePath: string): Promise<void> {
    const fullPath = path.join(this.storagePath, storagePath);

    if (fs.existsSync(fullPath)) {
      await fs.promises.unlink(fullPath);
      this.logger.log(`File deleted: ${storagePath}`);
    } else {
      this.logger.warn(`File not found for deletion: ${storagePath}`);
    }
  }

  fileExists(storagePath: string): boolean {
    const fullPath = path.join(this.storagePath, storagePath);
    return fs.existsSync(fullPath);
  }
}
